server-timing/defaults.php

```diff
@@ -211,3 +211,85 @@ static function( $acc, $query ) {
 	}
 }
 add_action( 'wp_loaded', 'perflab_register_default_server_timing_template_metrics' );
+
+/**
+ * Registers additional Server-Timing metrics as configured in the setting.
+ *
+ * These metrics should be registered as soon as possible. They can be added
+ * and modified in the "Tools > Server-Timing" screen.
+ *
+ * @since n.e.x.t
+ */
+function perflab_register_additional_server_timing_metrics_from_setting() {
+	$options = (array) get_option( PERFLAB_SERVER_TIMING_SETTING, array() );
+
+	/*
+	 * This closure measures performance of a hook (action or filter), as follows:
+	 *
+	 * 1. Add a hook callback at the minimum (i.e. earliest) priority possible.
+	 * 2. In that callback, register the metric for the hook, with a prefix of either "action" or "filter".
+	 * 3. Provide a measuring callback which captures the time span between beginning to end of the hook:
+	 *     1. Capture the current time immediately, i.e. at the earliest hook priority.
+	 *     2. Add another hook callback at the maximum (i.e. latest) priority possible.
+	 *     3. In that callback, capture the current time, leading the Server-Timing API to calculate the difference.
+	 *
+	 * Parameters to this closure are the hook name, what type of hook it is (either "action" or "filter", used as a
+	 * prefix for the registered metric), and the callback function to add the hook (either "add_action" or
+	 * "add_filter").
+	 */
+	$measure_hook = static function( $hook_name, $hook_type, $add_hook_func ) {
+		$metric_slug      = "{$hook_type}-{$hook_name}";
+		$measure_callback = static function( $metric ) use ( $hook_name, $add_hook_func ) {
+			$metric->measure_before();
+			call_user_func(
+				$add_hook_func,
+				$hook_name,
+				static function( $passthrough = null ) use ( $metric ) {
+					$metric->measure_after();
+					return $passthrough;
+				},
+				PHP_INT_MAX
+			);
+		};
+
+		call_user_func(
+			$add_hook_func,
+			$hook_name,
+			static function( $passthrough = null ) use ( $metric_slug, $measure_callback ) {
+				perflab_server_timing_register_metric(
+					$metric_slug,
+					array(
+						'measure_callback' => $measure_callback,
+						'access_cap'       => 'exist',
+					)
+				);
+				return $passthrough;
+			},
+			// phpcs:ignore PHPCompatibility.Constants.NewConstants.php_int_minFound
+			defined( 'PHP_INT_MIN' ) ? PHP_INT_MIN : ~PHP_INT_MAX
```

Comment:
> The callback can be added in the `all` hook _just in time_ so it is assured to be the very last last `PHP_INT_MAX` callback priority since `WP_Hook::add_filter()` will pushes onto the array.

Actually, it seems this is not relevant since the `PHP_INT_MAX`-priority hook is being added in the `PHP_INT_MIN` callback. Nevertheless, what remains is the possibility that another existing `PHP_INT_MIN` callback may not get measured?

Reply:
I modified the logic to use `all` in https://github.com/WordPress/performance/pull/784/commits/400c0a152b95dca374b9581e15ac4264e953ff9a.

Note that for now only hooks that run once are measured, or for hooks that run multiple times only the first occurrence is measured. That is something we can change in the future, but it would also require a few enhancements to even the underlying Server-Timing API, so it's out of scope for this. I have included logic though to ensure that hooks that run multiple times do not cause errors (e.g. due to duplicate metric registration), so it was also a good callout from that perspective.

Reply:
Humm, yeah. Being able to capture server timing for multiple hook runs seems important, doesn't it? 

Here are the hooks which are called more than once in a logged-out visit to the homepage on a vanilla site (via [Gist](https://gist.github.com/westonruter/fbf22063a74c7844c85ee2a8013b934e)):

Hook | Count
-- | --:
gettext | 1,507
gettext_default | 1,503
gettext_with_context | 1,231
gettext_with_context_default | 1,142
pre_wp_load_alloptions | 795
alloptions | 795
pre_option | 793
pre_option_stylesheet | 393
option_stylesheet | 393
stylesheet | 389
set_url_scheme | 293
should_load_separate_core_block_assets | 236
theme_root | 206
stylesheet_directory | 185
attribute_escape | 170
pre_option_siteurl | 162
option_siteurl | 162
site_url | 147
includes_url | 122
plugins_url | 113
sanitize_key | 110
block_type_metadata | 92
register_block_type_args | 92
block_type_metadata_settings | 91
gettext_with_context_twentytwentythree | 89
block_parser_class | 61
esc_html | 56
render_block | 51
safe_style_css | 51
pre_render_block | 48
render_block_data | 48
render_block_context | 48
pre_option_timezone_string | 42
option_timezone_string | 42
safecss_filter_attr_allow_css | 42
default_template_types | 41
sanitize_title | 35
clean_url | 33
register_post_type_args | 28
registered_post_type | 28
query | 25
pre_option_home | 22
option_home | 22
wp_register_sidebar_widget | 22
parse_tax_query | 22
home_url | 21
pre_option_gmt_offset | 21
option_gmt_offset | 21
pre_option_template | 19
option_template | 19
template | 19
is_post_type_viewable | 19
pre_option_permalink_structure | 15
option_permalink_structure | 15
register_taxonomy_args | 14
registered_taxonomy | 14
get_term | 14
block_default_classname | 14
no_texturize_tags | 14
no_texturize_shortcodes | 14
locale | 13
pre_option_sticky_posts | 13
option_sticky_posts | 13
kses_allowed_protocols | 12
sanitize_text_field | 12
content_url | 12
theme_root_uri | 12
stylesheet_directory_uri | 12
wp_lazy_loading_enabled | 12
template_directory | 11
parse_query | 11
pre_get_posts | 11
wp_allow_query_attachment_by_filename | 11
pre_option_comments_per_page | 11
option_comments_per_page | 11
posts_search | 11
get_terms_defaults | 11
parse_term_query | 11
pre_get_terms | 11
get_terms_args | 11
get_terms_orderby | 11
list_terms_exclusions | 11
get_meta_sql | 11
get_terms_fields | 11
terms_clauses | 11
terms_pre_query | 11
posts_where | 11
posts_join | 11
posts_where_paged | 11
posts_groupby | 11
posts_join_paged | 11
posts_orderby | 11
posts_distinct | 11
post_limits | 11
posts_fields | 11
posts_clauses | 11
posts_selection | 11
posts_where_request | 11
posts_groupby_request | 11
posts_join_request | 11
posts_orderby_request | 11
posts_distinct_request | 11
posts_fields_request | 11
post_limits_request | 11
posts_clauses_request | 11
posts_request | 11
posts_pre_query | 11
posts_results | 11
the_posts | 11
theme_file_uri | 11
theme_file_path | 10
admin_url | 10
wp_parse_str | 10
pre_kses | 10
pre_term_name | 9
pre_wp_theme_name | 9
get_wp_theme | 9
split_the_query | 9
pre_option_show_on_front | 9
option_show_on_front | 9
render_block_ | 9
pre_option_cron | 8
option_cron | 8
default_wp_template_part_areas | 8
get_post_metadata | 8
default_post_metadata | 8
get_the_terms | 8
pre_get_scheduled_event | 7
current_theme_supports-block-templates | 7
map_meta_cap | 6
user_has_cap | 6
pre_option_use_smilies | 6
option_use_smilies | 6
update_post_metadata_cache | 6
post_thumbnail_id | 6
get_user_metadata | 6
content_pagination | 6
the_title | 6
post_password_required | 6
pre_determine_locale | 5
sanitize_locale_name | 5
determine_locale | 5
pre_get_block_templates | 5
get_block_templates | 5
posts_request_ids | 5
get_category | 5
pre_option_blogname | 5
option_blogname | 5
render_block_core/group | 5
pre_option_blog_charset | 4
option_blog_charset | 4
rest_url | 4
ngettext | 4
ngettext_default | 4
pre_option_date_format | 4
option_date_format | 4
pre_site_option_can_compress_scripts | 4
default_site_option_can_compress_scripts | 4
pre_option_can_compress_scripts | 4
option_can_compress_scripts | 4
site_option_can_compress_scripts | 4
rest_url_prefix | 4
gettext_twentytwentythree | 4
perflab_server_timing_use_output_buffer | 4
show_admin_bar | 4
the_post | 4
pre_post_link | 4
post_link | 4
wp_date | 4
get_post_time | 4
get_the_date | 4
sanitize_html_class | 4
wp_resource_hints | 4
default_feed | 4
script_loader_src | 4
pre_option_theme_mods_twentytwentythree | 3
option_theme_mods_twentytwentythree | 3
nonce_user_logged_out | 3
nonce_life | 3
salt | 3
pre_option_blog_public | 3
option_blog_public | 3
wp_sitemaps_add_provider | 3
wp_theme_json_data_theme | 3
register_setting_args | 3
register_setting | 3
pre_option_html_type | 3
option_html_type | 3
wp_should_replace_insecure_home_url | 3
render_block_core/spacer | 3
render_block_core/paragraph | 3
bloginfo | 3
register_category_taxonomy_args | 2
taxonomy_labels_category | 2
registered_taxonomy_category | 2
register_post_tag_taxonomy_args | 2
taxonomy_labels_post_tag | 2
registered_taxonomy_post_tag | 2
register_nav_menu_taxonomy_args | 2
taxonomy_labels_nav_menu | 2
registered_taxonomy_nav_menu | 2
register_link_category_taxonomy_args | 2
taxonomy_labels_link_category | 2
registered_taxonomy_link_category | 2
register_post_format_taxonomy_args | 2
taxonomy_labels_post_format | 2
registered_taxonomy_post_format | 2
register_wp_theme_taxonomy_args | 2
taxonomy_labels_wp_theme | 2
registered_taxonomy_wp_theme | 2
register_wp_template_part_area_taxonomy_args | 2
taxonomy_labels_wp_template_part_area | 2
registered_taxonomy_wp_template_part_area | 2
register_post_post_type_args | 2
post_type_labels_post | 2
registered_post_type_post | 2
register_page_post_type_args | 2
post_type_labels_page | 2
registered_post_type_page | 2
register_attachment_post_type_args | 2
post_type_labels_attachment | 2
registered_post_type_attachment | 2
register_revision_post_type_args | 2
post_type_labels_revision | 2
registered_post_type_revision | 2
register_nav_menu_item_post_type_args | 2
post_type_labels_nav_menu_item | 2
registered_post_type_nav_menu_item | 2
register_custom_css_post_type_args | 2
post_type_labels_custom_css | 2
registered_post_type_custom_css | 2
register_customize_changeset_post_type_args | 2
post_type_labels_customize_changeset | 2
registered_post_type_customize_changeset | 2
register_oembed_cache_post_type_args | 2
post_type_labels_oembed_cache | 2
registered_post_type_oembed_cache | 2
register_user_request_post_type_args | 2
post_type_labels_user_request | 2
registered_post_type_user_request | 2
register_wp_block_post_type_args | 2
post_type_labels_wp_block | 2
registered_post_type_wp_block | 2
register_wp_template_post_type_args | 2
post_type_labels_wp_template | 2
registered_post_type_wp_template | 2
register_wp_template_part_post_type_args | 2
post_type_labels_wp_template_part | 2
registered_post_type_wp_template_part | 2
register_wp_global_styles_post_type_args | 2
post_type_labels_wp_global_styles | 2
registered_post_type_wp_global_styles | 2
register_wp_navigation_post_type_args | 2
post_type_labels_wp_navigation | 2
registered_post_type_wp_navigation | 2
plugin_loaded | 2
wp_video_extensions | 2
pre_option_perflab_modules_settings | 2
option_perflab_modules_settings | 2
perflab_active_modules | 2
pre_option_category_base | 2
option_category_base | 2
pre_option_tag_base | 2
option_tag_base | 2
pre_option_start_of_week | 2
option_start_of_week | 2
pre_option_time_format | 2
option_time_format | 2
wp_theme_json_data_default | 2
wp_theme_json_data_blocks | 2
wp_theme_json_data_user | 2
pre_option_rewrite_rules | 2
option_rewrite_rules | 2
get_terms | 2
wp_using_themes | 2
rest_queried_resource_route | 2
pre_get_shortlink | 2
get_shortlink | 2
loop_no_results | 2
render_block_core_template_part_file | 2
render_block_core/site-title | 2
render_block_core/template-part | 2
update_user_metadata_cache | 2
loop_start | 2
post_thumbnail_size | 2
post_thumbnail_html | 2
render_block_core/post-featured-image | 2
render_block_core/post-title | 2
get_the_excerpt | 2
excerpt_allowed_wrapper_blocks | 2
excerpt_allowed_blocks | 2
the_content | 2
excerpt_length | 2
excerpt_more | 2
wp_trim_words | 2
wp_trim_excerpt | 2
render_block_core/post-excerpt | 2
render_block_core/post-date | 2
render_block_core/null | 2
current_theme_supports-post-thumbnails | 2
has_post_thumbnail | 2
is_sticky | 2
post_class_taxonomies | 2
post_class | 2
render_block_core/column | 2
render_block_core/pattern | 2
wp_theme_json_get_style_nodes | 2
styles_inline_size_limit | 2
feed_content_type | 2
feed_link | 2
print_styles_array | 2
print_scripts_array | 2
script_loader_tag | 2
pre_option_wp_page_for_privacy_policy | 2
option_wp_page_for_privacy_policy | 2

Reply:
Sure, but we can always iterate. No need to implement everything we need in a single PR, that makes things extremely complex to review. FWIW, out of all the benchmarking we've been doing over the past few months I don't recall us ever having needed to benchmark a specific hook that runs multiple times. Having something that works for "single use" hooks available is most critical, and we can follow up with further enhancements after that.